microservices/notification/src/services/firebase/index.ts

```diff
@@ -0,0 +1,246 @@
+  | 'messaging/invalid-argument'
+  | 'messaging/server-unavailable';
+
+// Firebase has a limit of 500 messages per batch
+const BATCH_SIZE = 500;
+// Rate limit: 500 messages per second
+const RATE_LIMIT_DELAY = 1000;
+
+// Common error message
+const ERROR_NOT_INITIALIZED = 'Firebase service is not initialized';
+
+/**
+ * Service for handling Firebase Cloud Messaging operations
+ */
+class FirebaseService {
+  private static instance: FirebaseService;
+
+  private constructor() {
+    // Private constructor to enforce singleton pattern
+  }
+
+  /**
+   * Get singleton instance of FirebaseService
+   */
+  public static getInstance(): FirebaseService {
+    if (!FirebaseService.instance) {
+      FirebaseService.instance = new FirebaseService();
+    }
+
+    return FirebaseService.instance;
+  }
+
+  /**
+   * Send a batch of FCM messages
+   */
+  private async sendBatch(messages: Message[], messaging: Messaging): Promise<void> {
+    try {
+      await Promise.all(messages.map((message) => messaging.send(message)));
+    } catch (error) {
+      if (error instanceof Error) {
+        const errorMessage = error.message as FirebaseErrorCode;
+
+        switch (errorMessage) {
+          case 'messaging/invalid-registration-token':
+            throw new BaseException({
+              status: 422,
+              message: 'Invalid device token(s) provided',
+            });
+          case 'messaging/registration-token-not-registered':
+            throw new BaseException({
+              status: 422,
+              message: 'Device token(s) are no longer valid',
+            });
+          case 'messaging/invalid-argument':
+            throw new BaseException({
+              status: 422,
+              message: 'Invalid notification payload',
+            });
+          case 'messaging/server-unavailable':
+            throw new BaseException({
+              status: 503,
+              message: 'Firebase service is temporarily unavailable',
+            });
+          default:
+            throw new BaseException({
+              status: 500,
+              message: 'Failed to send notification batch',
+            });
+        }
+      }
+
+      throw new BaseException({
+        status: 500,
+        message: 'Failed to send notification batch',
+      });
+    }
+  }
+
+  /**
+   * Send push notification to multiple users
+   */
+  public async sendNotification(
+    userIds: string[],
+    title: string,
+    body: string,
+    data?: Record<string, string>,
+  ): Promise<void> {
+    if (!userIds.length) {
+      throw new BaseException({
+        status: 422,
+        message: 'No recipients specified',
+      });
+    }
+
+    const firebase = await FirebaseSdk();
+
+    let messaging: Messaging;
+
+    try {
+      messaging = firebase.messaging();
+    } catch (error) {
+      throw new BaseException({
+        status: 503,
+        message: ERROR_NOT_INITIALIZED,
+      });
+    }
+
+    const manager = getManager();
```

Comment:
better to pass repository via constructor
https://github.com/Lomray-Software/microservices/blob/staging/microservices/users/src/services/confirm/abstract.ts#L22